provisioning/service/tests/Config/AttestationMechanismTests.cs

```diff
@@ -11,21 +11,218 @@ namespace Microsoft.Azure.Devices.Provisioning.Service.Test
-        /* SRS_ATTESTATION_MECHANISM_21_013: [The constructor shall throw ArgumentException if the provided AttestationMechanismType is `tpm` but the tpm attestation is null.] */
-        /* SRS_ATTESTATION_MECHANISM_21_014: [If the provided AttestationMechanismType is `tpm`, the constructor shall store the provided tpm attestation.] */
-        /* SRS_ATTESTATION_MECHANISM_21_015: [The constructor shall throw ArgumentException if the provided AttestationMechanismType is `x509` but the x509 attestation is null.] */
+        //Requirement not testable.
+
+        /* SRS_ATTESTATION_MECHANISM_21_013: [The constructor shall throw ProvisioningServiceClientException if the provided AttestationMechanismType 
+                                                is `TPM` but the TPM attestation is null.] */
+        [TestMethod]
+        [TestCategory("DevService")]
+        public void AttestationMechanism_ConstructorJSON_ThrowsOnTypeTPMWithX509Attestation()
+        {
+            // arrange
+            string invalidJson =
+                "{\n" +
+                "   \"type\":\"tpm\",\n" +
+                "   \"x509\":{\n" +
+                "       \"signingCertificates\":{\n" +
+                "           \"primary\":{\n" +
+                "               \"info\": {\n" +
+                "                   \"subjectName\": \"CN=ROOT_00000000-0000-0000-0000-000000000000, OU=Azure IoT, O=MSFT, C=US\",\n" +
+                "                   \"sha1Thumbprint\": \"0000000000000000000000000000000000\",\n" +
+                "                   \"sha256Thumbprint\": \"" + ID + "\",\n" +
+                "                   \"issuerName\": \"CN=ROOT_00000000-0000-0000-0000-000000000000, OU=Azure IoT, O=MSFT, C=US\",\n" +
+                "                   \"notBeforeUtc\": \"2017-11-14T12:34:18Z\",\n" +
+                "                   \"notAfterUtc\": \"2017-11-20T12:34:18Z\",\n" +
+                "                   \"serialNumber\": \"000000000000000000\",\n" +
+                "                   \"version\": 3\n" +
+                "               }\n" +
+                "           }\n" +
+                "       }\n" +
+                "   }\n" +
+                "}";
+
+            // act - assert
+            TestAssert.Throws<ProvisioningServiceClientException>(() => Newtonsoft.Json.JsonConvert.DeserializeObject<AttestationMechanism>(invalidJson));
+        }
+
+        /* SRS_ATTESTATION_MECHANISM_21_014: [If the provided AttestationMechanismType is `TPM`, the constructor shall store the provided TPM attestation.] */
+        [TestMethod]
+        [TestCategory("DevService")]
+        public void AttestationMechanism_ConstructorJSON_SucceedForTPM()
+        {
+            // arrange
+            AttestationMechanism attestationMechanism = Newtonsoft.Json.JsonConvert.DeserializeObject<AttestationMechanism>(TPM_ATTESTATION_JSON);
```

Comment:
Here and everywhere else: avoid using fully qualified names and instead add `using` statements.

Reply:
Done